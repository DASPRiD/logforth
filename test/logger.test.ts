import { afterEach, beforeEach, expect, it, vi } from 'vitest';
import { LogEntry, Logger, LogLevel, Transport } from '../src/index.js';

class MockTransport implements Transport {
    public entries: LogEntry[] = [];

    public log(entry: LogEntry): void {
        this.entries.push(entry);
    }
}

const fakeTime = new Date(2000, 1, 1, 0, 0, 0, 0);

beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(fakeTime);
});

afterEach(() => {
    vi.useRealTimers();
});

it("should send log to transport", () => {
    const transport = new MockTransport();
    const logger = new Logger({ transport });
    logger.info("test");

    expect(transport.entries).toMatchSnapshot();
});

it("should send log with attributes to transport", () => {
    const transport = new MockTransport();
    const logger = new Logger({ transport });
    logger.info("test", {foo: "bar"});

    expect(transport.entries).toMatchSnapshot();
});

it("should merge default attributes", () => {
    const transport = new MockTransport();
    const logger = new Logger({ transport, attributes: {foo: "bar"} });
    logger.info("test", {baz: "bat"});

    expect(transport.entries).toMatchSnapshot();
});

it("should allow context specific attributes", async () => {
    const transport = new MockTransport();
    const logger = new Logger({ transport, attributes: {foo: "bar"} });

    await logger.withContext({
        baz: "bat",
    }, () => {
        logger.info("test", {another: "attribute"});
    })

    expect(transport.entries).toMatchSnapshot();
});

it("should omit entries below min level", async () => {
    const transport = new MockTransport();
    const logger = new Logger({ transport, minLevel: LogLevel.Info });
    logger.debug("foo");
    logger.info("foo");

    expect(transport.entries).toMatchSnapshot();
});
