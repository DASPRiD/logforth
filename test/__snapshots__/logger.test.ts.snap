// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow context specific attributes 1`] = `
[
  {
    "attributes": {
      "another": "attribute",
      "baz": "bat",
      "foo": "bar",
    },
    "level": 30,
    "message": "test",
    "time": 2000-01-31T23:00:00.000Z,
  },
]
`;

exports[`should merge default attributes 1`] = `
[
  {
    "attributes": {
      "baz": "bat",
      "foo": "bar",
    },
    "level": 30,
    "message": "test",
    "time": 2000-01-31T23:00:00.000Z,
  },
]
`;

exports[`should omit entries below min level 1`] = `
[
  {
    "attributes": {},
    "level": 30,
    "message": "foo",
    "time": 2000-01-31T23:00:00.000Z,
  },
]
`;

exports[`should send log to transport 1`] = `
[
  {
    "attributes": {},
    "level": 30,
    "message": "test",
    "time": 2000-01-31T23:00:00.000Z,
  },
]
`;

exports[`should send log with attributes to transport 1`] = `
[
  {
    "attributes": {
      "foo": "bar",
    },
    "level": 30,
    "message": "test",
    "time": 2000-01-31T23:00:00.000Z,
  },
]
`;
